File: UrlUtility.py
from __future__ import annotations

import re
import sys


def extract_url(url: str) -> str:
    url = url.replace("</a>", "")
    if "drive.google.com" in url:
        return gdrive_link_parse(url)
    elif "mega.nz" in url:
        return mega_link_parse(url)
    else:
        start = url.find("https:")
        return url[start:] if start != -1 else url


def gdrive_link_parse(url: str) -> str:
    start = url.find("https:")
    if start == -1:
        return ""
    m = re.search(r"(\?usp=sharing|\?usp=share_link|\?id=)", url)
    if m is not None:
        match m.group(1):
            case "?usp=sharing":
                end = m.span(1)[0] + len("?usp=sharing")
            case "?usp=share_link":
                end = m.span(1)[0] + len("?usp=share_link")
            case "?id=":
                end = m.span(1)[0] + len("?id=") + 33
            case _:
                print(f"Unable to identify gdrive url by parameter: {url}", file=sys.stderr)
                return ""
        
        return url[start:end] if len(url) >= end else ""
    
    m = re.search(r"(/folders/|/file/d/)", url)
    if m is not None:
        match m.group(1):
            case "/folders/":
                end = m.span(1)[0] + len("/folders/") + 33
            case "/file/d/":
                end = m.span(1)[0] + len("/file/d/") + 33
            case _:
                print(f"Unable to identify gdrive url by path: {url}", file=sys.stderr)
                return ""

        return url[start:end] if len(url) >= end else ""
    
    print(f"Unable to identify gdrive url: {url}", file=sys.stderr)
    return ""


def mega_link_parse(url: str) -> str:
    start = url.find("https:")
    if start == -1:
        return ""
    m = re.search(r"(/folder/|/#F!|/#!|/file/)", url)
    if m is None:
        print("Unable to identify mega url", file=sys.stderr)
        return ""
    match m.group(1):
        case "/folder/":
            end = m.span(1)[0] + len("/folder/") + 31
        case "/#F!":
            end = m.span(1)[0] + len("/#F!") + 31
        case "/#!":
            end = m.span(1)[0] + len("/#!") + 52
        case "/file/":
            end = m.span(1)[0] + len("/file/") + 52
        case _:
            print(f"Unable to identify mega url by path: {url}", file=sys.stderr)
            return ""

    return url[start:end] if len(url) >= end else ""


if __name__ == "__main__":
    pass

File: Sdk/Utility/UrlUtility.cs
using System.Text.RegularExpressions;
using Serilog;

namespace Sdk.Utility;

public static partial class UrlUtility
{
    private static readonly ILogger Logger = Log.ForContext(typeof(UrlUtility));
    
    public static string ExtractUrl(string url)
    {
        url = url.Replace("</a>", "");
        if (url.Contains("drive.google.com"))
        {
            return GDriveLinkParse(url);
        }

        if (url.Contains("mega.nz"))
        {
            return MegaLinkParse(url);
        }

        var start = url.IndexOf("https:", StringComparison.Ordinal);
        return start != -1 ? url[start..] : url;
    }

    private static string GDriveLinkParse(string url)
    {
        var start = url.IndexOf("https:", StringComparison.Ordinal);
        if (start == -1)
        {
            return "";
        }

        var m = GDriveLinkRegex1().Match(url);
        if (m.Success)
        {
            int end;
            switch (m.Groups[1].Value)
            {
                case "?usp=sharing":
                    end = m.Groups[1].Index + "?usp=sharing".Length;
                    break;
                case "?usp=share_link":
                    end = m.Groups[1].Index + "?usp=share_link".Length;
                    break;
                case "?id=":
                    end = m.Groups[1].Index + "?id=".Length + 33;
                    break;
                default:
                    Logger.Warning("Incorrect Match: {Url}", url);
                    return "";
            }

            return url.Length < end ? "" : url[start..end];
        }

        m = GDriveLinkRegex2().Match(url);
        if (m.Success)
        {
            int end;
            switch (m.Groups[1].Value)
            {
                case "/folders/":
                    end = m.Groups[1].Index + "/folders/".Length + 33;
                    break;
                case "/file/d/":
                    end = m.Groups[1].Index + "/file/d/".Length + 33;
                    break;
                default:
                    Logger.Warning("Incorrect Match: {Url}", url);
                    return "";
            }

            return url.Length < end ? "" : url[start..end];
        }

        Logger.Warning("Unrecognized GDrive url: {Url}", url);
        return "";
    }

    private static string MegaLinkParse(string url)
    {
        var start = url.IndexOf("https:", StringComparison.Ordinal);
        if (start == -1)
        {
            return "";
        }

        var m = MegaLinkRegex().Match(url);
        if (!m.Success)
        {
            Logger.Warning("Unrecognized Mega url: {Url}", url);
            return "";
        }

        int end;
        switch (m.Groups[1].Value)
        {
            case "/folder/":
                end = m.Groups[1].Index + "/folder/".Length + 31;
                break;
            case "/#F!":
                end = m.Groups[1].Index + "/#F!".Length + 31;
                break;
            case "/#!":
                end = m.Groups[1].Index + "/#!".Length + 52;
                break;
            case "/file/":
                end = m.Groups[1].Index + "/file/".Length + 52;
                break;
            default:
                Logger.Warning("Incorrect Match: {Url}", url);
                return "";
        }

        return url.Length < end ? "" : url[start..end];
    }
    
    public static string TruncateLongUrl(string url)
    {
        return string.Concat(url.AsSpan(0, 50), "...", url.AsSpan(url.Length - 49));
    }

    public static string GetUrlParameterValue(string url, string parameter)
    {
        return url.Split($"{parameter}=")[1].Split("&")[0];
    }
    
    [GeneratedRegex(@"(\?usp=sharing|\?usp=share_link|\?id=)")]
    private static partial Regex GDriveLinkRegex1();

    [GeneratedRegex(@"(/folders/|/file/d/)")]
    private static partial Regex GDriveLinkRegex2();

    [GeneratedRegex(@"(/folder/|/#F!|/#!|/file/)")]
    private static partial Regex MegaLinkRegex();
}